# CODING DISCUSSION 4 

### Submitted by HARSHINI(ht442)

In [263]:
# Importing packages
install.packages("ggcorrplot")
import numpy as np
import pandas as pd




NameError: name 'install' is not defined

### Reading Data 

In [250]:
# Reading the given five newspaper text files into 'alz','bbc','bbt','cnn' and 'fox' variables 
str1 = open("C:/Users/VIOLIN/Data/aljazeera-khashoggi.txt", "r",encoding = "UTF-8")
str2= open("C:/Users/VIOLIN/Data/bbc-khashoggi.txt", "r",encoding = "UTF-8")
str3 = open("C:/Users/VIOLIN/Data/breitbart-khashoggi.txt", "r",encoding = "UTF-8")
str4 = open("C:/Users/VIOLIN/Data/cnn-khashoggi.txt", "r",encoding = "UTF-8")
str5 = open("C:/Users/VIOLIN/Data/fox-khashoggi.txt", "r",encoding = "UTF-8")
swds = pd.read_csv("C:/Users/VIOLIN/Data/stop_words.csv")

# reading the data from the loaded txt files
alz = str1.read()
bbc = str2.read()
bbt = str3.read()
cnn = str4.read()
fox = str5.read()

# Creating list of Words from the stop words csv file 

swds_list = swds["word"].values 

#### Filtering  the newspaper articles for required words
#### Defining tokenize function to break the strings into bucket of words. 

In [251]:
def tokenize(text=None):
    """
    The defined Tokenize function removes punctuations marks and any other non letter details of the string, stopwords(from the csv file) and returns 
    string as a uniform list of words in lower case 
    
    """
    #Converting the text into Lower case
    text = text.lower()
    #Replacing punctuation marks and other non letter details from the string with single quotes
    text = text.replace('.','')
    text = text.replace(',','')
    text = text.replace('(','')
    text = text.replace(')','')
    text = text.replace('"','')
    text = text.replace('-','')
    text = text.replace('_','')
    text = text.replace("“", "")
    text = text.replace("”", "")
    text = text.replace('[','')
    text = text.replace(']','')
    text = text.replace("\'s",'')
    text = text.replace("?",'')
    text = text.replace("\'",'')
    text = text.replace('{','')
    text = text.replace('}','')

    #Splitting the string into blocks of words

    list1 = text.split()
    # Removing stop words from the function output
    list2 = [i for i in list1 if i not in swds_list]

    return list2









In [252]:
## Applying the tokenize function to all the five strings alz,bbc,bbt,cnn and fox to break each string into its word blocks and Checking the Output
alz_tkn = tokenize(alz)
bbc_tkn = tokenize(bbc)
bbt_tkn = tokenize(bbt)
cnn_tkn = tokenize(cnn)
fox_tkn = tokenize(fox)

print(alz_tkn) # checking output


['turkey', 'istanbul', 'turkish', 'president', 'recep', 'tayyip', 'erdogan', 'murder', 'journalist', 'jamal', 'khashoggi', 'kingdom', 'consulate', 'istanbul', 'planned', 'saudi', 'officials', 'days', 'advance', 'addressing', 'legislators', 'justice', 'development', 'party', 'ak', 'party', 'tuesday', 'erdogan', 'detailed', 'khashoggi', 'disappearance', 'murder', 'stopped', 'short', 'accusing', 'saudi', 'royals', 'savage', 'killing', 'caused', 'global', 'outrage', 'september', '28', 'khashoggi', 'arrived', 'saudi', 'arabian', 'consulate', 'sort', 'wedding', 'paperwork', 'erdogan', 'speech', 'turkish', 'parliament', 'capital', 'ankara', 'time', 'saudi', 'arabian', 'officials', 'started', 'plan', 'roadmap', 'murder', 'added', 'saudi', 'officials', 'left', 'turkey', 'travelled', 'saudi', 'arabia', 'indicating', 'planned', 'murder', 'khashoggi', '59', 'washington', 'post', 'columnist', 'critic', 'powerful', 'saudi', 'crown', 'prince', 'mohammed', 'bin', 'salman', 'disappeared', 'entering', '

#### Creating a Document Term Matrix

In [253]:
def Dict(text1=None):
    """
    The defined Dict Function creates a document term matrix(frequency of each word in the document in matrix form)
    """
    # Create an Empty dictionary
    
    di = {}
     # Counting the frequency of words using if else statement
    for word in tokenize(text1):
        
        if word in di:
            di[word][0] += 1 #to count the frequency of existing words
        else:
            di[word] = [1]
    return pd.DataFrame(di)
            
    #Creating the output as dataframe
    return pd.DataFrame(di,ignore_index=True)

def DTM(text2):
    """
    The DTM function generates a Document Term Matrix in the form of a dataframe for all the news articles
    """
    # Create a dataframe
    DTM = pd.DataFrame()
    
    for text in text2:
        entry = Dict(text)
        #Row bind # Binding multiple dataframes
        DTM = DTM.append(pd.DataFrame(entry),ignore_index=True,sort=True)
    
    DTM.fillna(0, inplace=True) #to Fill empty boxes of the dataframe with zeros to avoid confusion
    
    return DTM
    

In [254]:
Document_term_matrix = DTM([alz,bbc,bbt,cnn,fox])#Creating the Document term matrix for our newspaper articles

In [255]:
Document_term_matrix #Viewing output

$50bn    1   11   12   15  15member   18  1:08    2   28  ...  widely  \
0    0.0  0.0  0.0  0.0  0.0       1.0  2.0   0.0  2.0  1.0  ...     0.0   
1    1.0  0.0  0.0  1.0  0.0       0.0  2.0   0.0  1.0  0.0  ...     1.0   
2    0.0  0.0  0.0  0.0  1.0       0.0  0.0   0.0  0.0  0.0  ...     0.0   
3    0.0  0.0  0.0  0.0  0.0       0.0  1.0   0.0  0.0  0.0  ...     0.0   
4    0.0  1.0  1.0  0.0  1.0       0.0  1.0   1.0  1.0  0.0  ...     0.0   

   withheld  woods  world  worse  writer  yalova  yelova  £385bn    —  
0       0.0    0.0    0.0    0.0     0.0     0.0     0.0     0.0  0.0  
1       0.0    0.0    1.0    0.0     0.0     0.0     0.0     1.0  0.0  
2       1.0    0.0    1.0    0.0     0.0     0.0     0.0     0.0  0.0  
3       0.0    0.0    0.0    0.0     0.0     1.0     0.0     0.0  0.0  
4       0.0    1.0    0.0    1.0     2.0     0.0     1.0     0.0  3.0  

[5 rows x 655 columns]

In [256]:
 Document_term_matrix = Document_term_matrix .rename(index={0:"aljazeera", 1:"bbc", 2:"breitbart", 3:"cnn", 4:"fox"}) # Adding index details for more readability

In [257]:
Document_term_matrix #Viewing index added output

$50bn    1   11   12   15  15member   18  1:08    2   28  ...  \
aljazeera    0.0  0.0  0.0  0.0  0.0       1.0  2.0   0.0  2.0  1.0  ...   
bbc          1.0  0.0  0.0  1.0  0.0       0.0  2.0   0.0  1.0  0.0  ...   
breitbart    0.0  0.0  0.0  0.0  1.0       0.0  0.0   0.0  0.0  0.0  ...   
cnn          0.0  0.0  0.0  0.0  0.0       0.0  1.0   0.0  0.0  0.0  ...   
fox          0.0  1.0  1.0  0.0  1.0       0.0  1.0   1.0  1.0  0.0  ...   

           widely  withheld  woods  world  worse  writer  yalova  yelova  \
aljazeera     0.0       0.0    0.0    0.0    0.0     0.0     0.0     0.0   
bbc           1.0       0.0    0.0    1.0    0.0     0.0     0.0     0.0   
breitbart     0.0       1.0    0.0    1.0    0.0     0.0     0.0     0.0   
cnn           0.0       0.0    0.0    0.0    0.0     0.0     1.0     0.0   
fox           0.0       0.0    1.0    0.0    1.0     2.0     0.0     1.0   

           £385bn    —  
aljazeera     0.0  0.0  
bbc           1.0  0.0  
breitbart     0.0  0.0  
cnn           0.0  0.0  
fox           0.0  3.0  

[5 rows x 655 columns]

In [258]:
# Defining Cosine function
def cosine(a,b):
    """
    The defined cosine function calculates the cosine value of the angle b/w the variables 'a','b' by using vector algebra concepts 
    of cos theta as dot product of (a,b) divided by magnitudes of 'a' &'b'
    
    """
    # using numpy package and  vector angle formula from math, define cos 
    cos = np.dot(a,b)/(np.sqrt(np.dot(a,a)) * np.sqrt(np.dot(b,b))) # 
    return cos #returns output of cosine function

In [259]:
# Creates an Correlation matrix to see the similarity of two articles by comparing their words
def correl_matrix(df):
    """
    The defined sm function produces a correlation matrix, by using cosine function defined earlier and numerical inputs 
    generated in Document_term_matrix and helps us identify the similarity of two vector variables. 
    
    """
   
    df_CM = pd.DataFrame(columns = df.index.values, index = df.index.values) #creating the dataframe
    
    for col in df_CM.columns:
        for row in df_CM.index.values:
            df_CM.loc[row, col] = cosine(df.loc[row].values, df.loc[col].values) #generating a dataframe of cosine values
            
    return(df_CM)# for printing output of the function

In [260]:
#Calling correl_matrix function for the output Document_term_matrix

In [262]:
correl_matrix(Document_term_matrix) #printing output of correlation values/correlation matrix

aljazeera       bbc breitbart       cnn       fox
aljazeera         1  0.704952  0.588124  0.590484  0.709046
bbc        0.704952         1  0.588177  0.556742  0.662161
breitbart  0.588124  0.588177         1  0.374829   0.54694
cnn        0.590484  0.556742  0.374829         1  0.558802
fox        0.709046  0.662161   0.54694  0.558802         1

##### Data Visualization

In [ ]:
# Scatter plot with square markers


plt.figure(figsize=(10, 10))
corrplot(data.corr())

### Output Analysis

##### Aljazeera,BBC and Fox News texts were closely correlated(0.7).
##### Breitbart and CNN have the lowest similarity values with each other(0.37). 
##### Also, Breitbart and CNN has a similar similarity levels with the other three texts(in the range of 0.5-0.6)
